server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java

```diff
@@ -100,7 +100,10 @@ protected Runnable cancelExistingRecoveryForBetterMatch(
         Metadata metadata = allocation.metadata();
         RoutingNodes routingNodes = allocation.routingNodes();
         ShardRouting primaryShard = allocation.routingNodes().activePrimary(shard.shardId());
-        assert primaryShard != null : "the replica shard can be allocated on at least one node, so there must be an active primary";
+        if (primaryShard == null) {
+            logger.trace("{}: no active primary shard found or allocated, letting actual allocation figure it out", shard);
+            return null;
+        }
```

Comment:
Above we already return `null` for the same NPE check
<img width="690" alt="Screenshot 2024-06-18 at 12 36 26" src="https://github.com/opensearch-project/OpenSearch/assets/23433684/f8d48859-6ad4-4527-a533-329086f922e8">

Reply:
As I understand one is about the primary shard being in an assigned state the other is whether or not the physical data it hosts is present. Both IMO are different